Implement "hab job promote"

raskchanky opened the pull request:
This introduces the `hab job promote` command, which gives the ability to take a job group ID and a channel name and promote every successfully built project inside that job group to the specified channel.

Part of this change is a new API endpoint in `builder-api` which does this promotion in one shot, instead of requiring N+1 HTTP requests for individually promoting every successfully built project inside a job group.  `builder-scheduler` was refactored to utilize this new endpoint.

![tenor-51992790](https://user-images.githubusercontent.com/947/28941816-12114b78-784e-11e7-9cac-2be8e4bcef17.gif)

christophermaier commented on components/builder-api/doc/api.raml:
```diff
@@ -267,6 +267,21 @@ securitySchemes:
+                            404:
+                                description: Group not found
+                            500:
+                                description: Internal server error
```
Based on the description, there are some projects (namely failed ones) in a group that would not get promoted. Can / should we give any indication of which these are? What if the group itself failed (e.g., some package toward the root of the dependency tree, and not a leaf, failed); what error would be returned then? What happens if you promote to a channel that the packages are already in?

raskchanky replied:
This was mostly just a port of the functionality as it existed in the scheduler, but you bring up some good points.

* If any of the projects didn't succeed, we can return those and indicate that they failed.
* If the group itself failed, we should indicate that in the return value as well.

For both of the above items, keeping the status as 200 seems appropriate, no?

If you promote to a channel that the packages are already in, nothing happens.  It goes through like it normally would.

raskchanky replied:
Scratch that - I'm going to return a 403 instead of 200 if the group itself failed, since the request can't go anywhere from there.

chefsalim replied:
Couple of notes:
1. We don't 'Fail' groups (except in very rare circumstances) if a package build fails. Mostly we will skip any packages (and their deps) that fail, but still complete the group.  
2. If any build failures happen, we will then use the cli tool for promotion.  So we should NOT use failed projects as a reason to not do the promote.  But it would be fine to warn that some projects didn't succeed and did not get promoted.
3. If the group itself got marked as Failed, then it's valid to return a 403 I think.

raskchanky replied:
@chefsalim That's in line with the changes I'm currently making.

christophermaier commented on components/builder-api/src/http/handlers.rs:
```diff
@@ -113,6 +115,56 @@ pub fn github_authenticate(req: &mut Request) -> IronResult<Response> {
+            do_channel_creation(req, ident.get_origin(), &channel, session_id)?;
+        }
+
+        do_promotion(req, ident, &channel, session_id)?;
```
It seems that we could get into a state here where some of the packages get successfully promoted, then something fails for whatever reason, and the remainder of the packages get left behind.

It'd be nice if we could instead to the spiritual equivalent of passing a list of identifiers into a SQL function somewhere and doing it all in a single transactional operation.

raskchanky replied:
Will do

chefsalim replied:
Definitely agree we should switch over to ZMQ instead of http in the scheduler.  
The common ZMQ helper code to do the promotion/channel creation could be moved into builder_core, so that both the scheduler and the depot can call it.

raskchanky commented on components/builder-depot/src/server.rs:
```diff
@@ -1446,39 +1446,21 @@ fn create_channel(req: &mut Request) -> IronResult<Response> {
+    };
+    channel = match params.find("channel") {
+        Some(channel) => channel.to_string(),
+        _ => return Ok(Response::with(status::BadRequest)),
```
Good call.

raskchanky commented on components/builder-depot/src/server.rs:
```diff
@@ -1978,6 +1915,88 @@ fn target_from_headers(user_agent_header: &UserAgent) -> result::Result<PackageT
+            match err.get_code() {
+                ErrCode::ENTITY_NOT_FOUND => Ok(Response::with((status::NotFound))),
+                _ => {
+                    error!("promote_package:1, err={:?}", err);
```
I'm renumbering all of them, good catch.

chefsalim commented on components/builder-api/src/http/handlers.rs:
```diff
@@ -113,6 +115,56 @@ pub fn github_authenticate(req: &mut Request) -> IronResult<Response> {
 }
 
+pub fn job_group_promote(req: &mut Request) -> IronResult<Response> {
+    // JB TODO: eliminate the need to clone the params and session - HI SALIM =)
```
LOL :)

chefsalim commented:
Definitely agree on the http-to-ZMQ switch-over for the scheduler.  It should be pretty straightforward to share the boilerplate code via builder_core for example, and will simplify some other things as well.  (We can look at the builder-worker http calls next as well).

raskchanky commented:
@christophermaier This is ready for a fresh review.  I believe I've addressed all of your feedback.

raskchanky commented:
I'll also leave a note here about my last commit where I removed some tests from builder-depot.  I'm of the opinion that mock based testing is brittle and not the best way to test.  I think builder-depot is best tested through our integration test suite, located in the `test` subdirectory at the root of the project.  I have a branch where I'm working through some of the removed tests there, but given how long this particular PR has been outstanding, I felt it was better to submit this now with the promise that more tests are coming, rather than hold this up and wait on those integration tests.

christophermaier commented on components/builder-core/src/api.rs:
```diff
@@ -0,0 +1,275 @@
+use error::{Error, Result};
+
+pub fn get_origin<T: ToString>(origin: T) -> result::Result<Option<Origin>, NetError> {
+    let mut conn = Broker::connect().unwrap();
```
I guess you decided against piping a `Broker` into the function?

raskchanky replied:
I tried to crack this nut about 20 different ways, and all of them either failed or felt terrible.  In the end, I chatted with @reset about how we were testing this stuff and decided that writing integration tests for it, using our existing integration test setup, was a better way to go than trying different ways of unit testing this.  Those integration tests are not part of this PR, as some of them proved very challenging to get working, and I didn't want this PR to wilt on the vine.

christophermaier commented on components/builder-core/src/api.rs:
```diff
@@ -0,0 +1,275 @@
+        }
+    };
+
+    let mut conn = Broker::connect().unwrap();
```
What's the cost of setting up a Broker connection? Wondering if piping a Broker through and reusing it might be nicer from that perspective, too.

raskchanky replied:
The cost is super low, just connecting a zmq socket.

christophermaier commented on components/builder-core/src/api.rs:
```diff
@@ -0,0 +1,275 @@
+    }
+}
+
+pub fn create_channel(origin: &str, channel: &str, session_id: u64) -> Result<(OriginChannel)> {
```
Is this really supposed to be `Result<(OriginChannel)>` instead of `Result<OriginChannel>`?

raskchanky replied:
Nope, typo.  Turns out it compiles either way 😁

christophermaier commented on components/builder-core/src/api.rs:
```diff
@@ -0,0 +1,275 @@
+
+    match conn.route::<OriginChannelGet, OriginChannel>(&channel_req) {
+        Ok(origin_channel) => {
+            if !check_origin_access(session_id, ident.get_origin())
```
Better to check this before making unnecessary network requests.

raskchanky replied:
Good call

christophermaier commented on components/builder-core/src/api.rs:
```diff
@@ -0,0 +1,275 @@
+        if project.get_state() == ProjectState::Success {
+            let ident = OriginPackageIdent::from_str(project.get_ident()).unwrap();
+
+            if channel != "stable" {
```
Better to refer to [STABLE_CHANNEL](https://github.com/habitat-sh/habitat/blob/46baddf1e2ee57d921913e383d95ac0f8ac45b66/components/core/src/channel.rs#L18) here

raskchanky replied:
Will do

christophermaier commented on components/builder-core/src/api.rs:
```diff
@@ -0,0 +1,275 @@
+            let ident = OriginPackageIdent::from_str(project.get_ident()).unwrap();
+
+            if channel != "stable" {
+                match create_channel(ident.get_origin(), channel, session_id) {
```
This is going to get executed for each package; can we move it up?

Also, I believe we handle channel creation at different places in other flows (e.g., up at the CLI layer). Might be worth taking a survey and seeing if there's a cause to consolidate or rethink where we're doing various operations.

raskchanky replied:
We can't move it from where it's at because the origin to create the channel for comes from the current package in the iteration, and there's no way to guarantee that every package will have the same origin.

raskchanky commented on components/http-client/src/util.rs:
```diff
@@ -0,0 +1,32 @@
+
+use error::{Error, Result};
+
+pub fn decoded_response<T>(mut response: Response) -> Result<T>
```
Consuming the `Response` is ok.  We only need the decoded thing out of it.  `T` is any type that implements `serde::de::DeserializeOwned`, as shown in the `where` clause.  In practice, this means that we pass a `Response` to this function and get back a deserialized thing from whatever JSON was present in this `Response`.

raskchanky commented on components/builder-api/src/http/handlers.rs:
```diff
@@ -530,7 +530,7 @@ pub fn project_update(req: &mut Request) -> IronResult<Response> {
             (body.github.organization, body.github.repo)
         }
-        _ => return Ok(Response::with(status::UnprocessableEntity)),
+        Err(_) => return Ok(Response::with(status::UnprocessableEntity)),
```
Will do

raskchanky commented on components/builder-api/src/http/handlers.rs:
```diff
@@ -145,24 +147,39 @@ pub fn job_group_promote(req: &mut Request) -> IronResult<Response> {
+    // This only makes sense if the group is complete. If the group isn't complete, return now and
+    // let the user know.
+    if group.get_state() != GroupState::Complete {
+        return Ok(Response::with((status::Forbidden, "rg:jgp:1")));
```
I have no idea what `rg` stands for.  I cargo-culted it from elsewhere in this file, but I figured it was good to be consistent 😁

raskchanky commented on components/builder-api/src/http/handlers.rs:
```diff
@@ -145,24 +147,39 @@ pub fn job_group_promote(req: &mut Request) -> IronResult<Response> {
+
+    for project in group.get_projects().into_iter() {
+        if project.get_state() == ProjectState::Success {
+            let ident = OriginPackageIdent::from_str(project.get_ident()).unwrap();
```
I don't see a `parse` method for `OriginPackageIdent`.  I did see a few places that used `from_str` though.

raskchanky commented on components/builder-api/src/http/handlers.rs:
```diff
@@ -145,24 +147,39 @@ pub fn job_group_promote(req: &mut Request) -> IronResult<Response> {
+        if project.get_state() == ProjectState::Success {
+            let ident = OriginPackageIdent::from_str(project.get_ident()).unwrap();
+
+            if &channel != "stable" {
```
Will do

raskchanky commented on components/builder-api-client/src/lib.rs:
```diff
@@ -35,8 +36,10 @@ pub use error::{Error, Result};
 use std::io::Read;
 use std::path::Path;
 
+use builder_core::data_structures::UnsuccessfulJobGroupPromote;
```
My intent, as I laid out in a comment in that file, was to have a place to put common data structures that are required by both the client and the server, e.g. the server generates a JSON response that represents some data structure and the client needs to parse that and re-construct it.  `data_structures` felt like it accurately represented the intent to me, but I'm certainly open to suggestions.

raskchanky commented on components/builder-api-client/src/lib.rs:
```diff
@@ -148,16 +151,34 @@ impl Client {
+            Ok(value) => Ok(value.projects),
+            Err(Error::HabitatHttpClient(hab_http::Error::Json(e))) => {
+                if e.is_eof() {
+                    return Ok(Vec::new());
```
This error is specific to the way that serde returns things when it parses JSON that is an empty string, e.g. the server returned an empty response body.  The server is returning either nothing, indicating success, or a list of projects that failed to promote.  I guess I could change the server to return an empty list of projects that failed to promote, but imo, that makes the server API more awkward.  If a request succeeds, we don't return an empty list of errors - we just return nothing.  Either way, something is a little awkward, and imo, this strikes the right balance as it stands.

raskchanky commented on components/builder-api/src/http/handlers.rs:
```diff
@@ -145,24 +147,39 @@ pub fn job_group_promote(req: &mut Request) -> IronResult<Response> {
-        do_promotion(req, ident, &channel, session_id)?;
+            do_promotion(req, ident, &channel, session_id)?;
+        } else {
+            projects.push(project.get_ident().to_string());
```
Sure

raskchanky commented on components/hab/src/error.rs:
```diff
@@ -127,6 +128,13 @@ impl fmt::Display for Error {
             Error::IO(ref err) => format!("{}", err),
+            Error::JobGroupPromote(ref id, ref c) => {
+                format!(
+                    "Failed to promote job group ID {} to channel {} because the job group is not complete.",
```
Hmm, the user already has to know what a job group and a channel is to use this CLI command.  We've decided against using the term "project" in any user facing text, to avoid confusion with packages. There are 4 possible reasons that a project might not've succeeded, but listing them all out here feels a bit overly verbose.

christophermaier replied:
OK, so maybe say "because some packages have not yet been built".

What are the four reasons a project might not have succeeded? Are those listed anywhere that a user could access them?

raskchanky replied:
There is an enum in the `scheduler.proto` file that lists out all the states a project could be in.  None of them are exposed to users.  I'll update this text.

raskchanky commented on components/builder-originsrv/src/server/handlers.rs:
```diff
@@ -646,6 +646,23 @@ pub fn origin_channel_package_list(
+        Ok(()) => req.reply_complete(sock, &net::NetOk::new())?,
+        Err(err) => {
+            error!("OriginPackageGroupPromote, err={:?}", err);
+            let err = net::err(ErrCode::DATA_STORE, "vt:origin-package-group-promote:1");
```
In general, I ensure that I'm being internally consistent with whatever module I'm working in. If the module in question spells everything out, I spell everything out. If the module in question abbreviates things, I abbreviate them. But yes, we are lacking consistency across the entire project.

raskchanky commented on components/builder-originsrv/tests/data_store/mod.rs:
```diff
@@ -1714,6 +1714,104 @@ fn get_origin_channel() {
+    ident1.set_origin("core".to_string());
+    ident1.set_name("cacerts".to_string());
+    ident1.set_version("2017.01.17".to_string());
+    ident1.set_release("20170209064044".to_string());
```
You have a good point, but this test code reads like all of the other test code in this file.  At some point, it might be worth refactoring all of these tests to use helper functions and clean up the verbosity, but as it is, someone reading through this test code today won't have any trouble following my test because it's like all of the other tests. I fear that choosing this moment to refactor one test will be jarring and make this test harder to follow.

raskchanky commented on components/builder-core/src/data_structures.rs:
```diff
@@ -0,0 +1,23 @@
+#[derive(Debug, Serialize, Deserialize)]
+pub struct UnsuccessfulJobGroupPromote {
+    pub group_id: u64,
+    pub projects: Vec<String>,
```
Will do

raskchanky commented on components/hab/src/command/job/promote.rs:
```diff
@@ -39,12 +41,35 @@ pub fn start(
+                ui.status(
+                    Status::Promoted,
+                    format!(
+                        "job group {} to channel {}, but the following projects failed to promote:",
```
Will do

raskchanky commented:
@christophermaier I believe I've addressed all of your comments, either with a comment of my own or a code change.  I looked through all of the outdated comments as well, in an effort to make sure I didn't overlook anything, but it's possible that something got lost in the shuffle.  It was pretty confusing sorting out which comments were on old code and which weren't.  😛 

I've also squashed all the commits down, which should hopefully make this a little easier to look at.

reset commented:
@raskchanky I will get you a review tonight or early tomorrow :)

raskchanky commented:
@reset Thanks man 🎉

christophermaier commented on components/builder-core/src/api.rs:
```diff
@@ -0,0 +1,277 @@
+    if let Some(core_projects) = origin_map.remove("core") {
+        let promote_result = do_group_promotion(channel, core_projects, "core", session_id);
+        if promote_result.is_err() {
+            return promote_result;
```
Bumping [this comment](https://github.com/habitat-sh/habitat/pull/2866#discussion_r133004927)... I think it was attached to older code (sorry! ☹️ )

raskchanky replied:
No, it shouldn't.  I'll correct that.

christophermaier commented on components/builder-core/src/api.rs:
```diff
@@ -0,0 +1,277 @@
+    if failed_projects.is_empty() {
+        Ok(())
+    } else {
+        let ujgp = UnsuccessfulJobGroupPromote {
```
Is it "unsuccessful" or "partially successful"? What does it mean for some projects to not get promoted? What remediation strategies (if any) does the user have in this situation? How should the user respond when they get this result?

raskchanky replied:
When I wrote this, I called it "unsuccessful" because the entire job group wasn't promoted successfully.  I can rename it to reflect the partial failure/partial success.  I changed the error message that `hab` returns to address your other questions.

christophermaier commented on components/builder-api-client/src/lib.rs:
```diff
@@ -139,10 +142,45 @@ impl Client {
+            return Err(err_from_response(res));
+        }
+
+        match decoded_response::<UnsuccessfulJobGroupPromote>(res)
```
Reviving [this thread](https://github.com/habitat-sh/habitat/pull/2866#discussion_r132986922)... I feel like this response could be structured differently so that this code is not as confusing. Maybe instead of calling it `UnsuccessfulJobGroupPromote`, it's just a `JobGroupPromoteResult` (or whatever). If no projects failed to promote, yay, that's a successful promotion. If some failed, then we create an appropriate error. That way, the flow is a bit more clear, and you don't have things like errors-that-aren't-really-errors.

raskchanky replied:
I'm happy to rename the struct if you think it would make things clearer, but that won't affect the error handling below.

By abstracting away the parsing of response bodies into `decoded_response`, we've lost the ability to have fine grained control over what to do with empty response bodies, since that method returns something that can be deserialized by serde.  Empty response bodies are perfectly normal for an HTTP API to have, but when you try and parse an empty response body with serde, you get an error, which we handle below.

I'm of the opinion that it's better to have your public facing HTTP API be correct and follow the principle of least surprise, and handle any oddities in your client libraries.  That's how this is setup right now.  We can change it so that the client code is more straightforward from an error handling perspective, but doing so will require making our HTTP API less intuitive, in my opinion.

In this particular case, there's just a wrinkle in how we handle success and failure, and I think it's best if it's not exposed to end users.

christophermaier replied:
Can `decoded_response` be tweaked to handle empty responses? i.e., can we push _that_ behavior further down so every caller of `decoded_response` doesn't have to handle it themselves?

I take your point about not exposing weird internal quirks via the external API, though. I'm fine with adding some explanatory comments to the code as it is. It still feels not-quite-right to me, but we can revise our approach as needed in the future.

raskchanky replied:
I'll add the comments for now and we can tweak it in the future if we need to.

christophermaier commented on components/hab/src/command/job/promote.rs:
```diff
@@ -0,0 +1,79 @@
+                }
+                println!("");
+                println!(
+                    "It's possible this failure was transient. You can try re-running this job group. You can also run build jobs for each of these packages individually."
```
Re-running or re-promoting?

Is there any way of knowing if this failure _wasn't_ transient? What would a user do in that case? How would a partial promotion of packages in a group affect their workflow / infrastructure / etc using these packages?

raskchanky replied:
Re-running, as in submitting the job to the scheduler again.  No, there's no way of knowing if the failure wasn't transient.  As far as I can tell, re-running is the best option.  I'm open to suggestions on better ways to word this.

christophermaier commented on components/hab/src/command/job/promote.rs:
```diff
@@ -0,0 +1,79 @@
+                ui.status(
+                    Status::Promoted,
+                    format!(
+                        "job group {} to channel {}, but the following packages failed to promote:",
```
Now that I read this message again, I wonder if it would be beneficial to also indicate which packages in the group did get promoted successfully.

raskchanky replied:
In my opinion, those are better left out.  The messaging is already feeling a bit Wall-of-Texty and there's nothing actionable around the packages that succeeded.  If you need to re-run something, all you really care about is the stuff that failed.

raskchanky commented on components/builder-core/src/api.rs:
```diff
@@ -0,0 +1,277 @@
+    // currently have.
+    for project in group.get_projects().into_iter() {
+        if project.get_state() == ProjectState::Success {
+            let ident = OriginPackageIdent::from_str(project.get_ident()).unwrap();
```
I'm probably missing something, but `from_str` is being called on `OriginPackageIdent`, not `String/str`.  I can't see anywhere where `parse` is defined on `OriginPackageIdent`.

christophermaier replied:
To call `from_str` you need a string, which is what `project.get_ident()` gives you; you call `parse` on that.

(To be clear, you're getting the same result. I'm just pointing out things from an idiomatic Rust perspective; that's all.)

raskchanky replied:
I think I'd prefer to leave this as is.  I've seen other places in the code that use `OriginPackageIdent::from_str` and it feels better to me to have this be consistent.

raskchanky commented on components/builder-core/src/api.rs:
```diff
@@ -0,0 +1,277 @@
+            let ident = OriginPackageIdent::from_str(project.get_ident()).unwrap();
+
+            if channel != STABLE_CHANNEL {
+                match create_channel(ident.get_origin(), channel, session_id) {
```
No, because it's possible that we have projects from different origins.

christophermaier replied:
Right, but after you've inserted things into the origin map, you're going through each origin to promote all the packages from that origin. If you do the channel creation _then_, you'd only be doing it once per origin, instead of once per project.

raskchanky replied:
Ah, I didn't get what you were saying the first time.  Sure, I will move it.

raskchanky commented on components/builder-api/src/http/handlers.rs:
```diff
@@ -163,18 +172,37 @@ pub fn job_group_promote(req: &mut Request) -> IronResult<Response> {
+    for (origin, projects) in origin_map.iter() {
+        let promote_result = do_group_promotion(&channel, projects.to_vec(), &origin);
+        if !promote_result.is_ok() {
+            return promote_result;
```
The more I think about how to solve this, the messier it feels.  Right now, pretty much everywhere in our codebase that I've seen, when an operation returns an error, we immediately return that error.  The returned error bubbles up, eventually reaching the very top of the call chain in `main.rs` where the process terminates with a non-zero exit code.  This follows that same pattern.

In thinking through how this would work if we didn't do that, and instead continued on if there was an error, we'd need some way of keeping track of the errors as they occurred and storing them for later use.  Then, once we had finished trying to promote everything, we'd have to somehow return this collection of errors and then figure out how to elegantly display multiple errors to the user.  Even if we solved the problem of displaying those errors to the user in a way that made sense, how would these multiple errors interact with Rust's notion of `Result` which accounts for a single success and single error value.  What gets bubbled back up to `main.rs`?

Maybe I'm just thinking about this wrong, but going this route feels like it opens a Pandora's Box of issues.

christophermaier replied:
I think it's OK that it feels messy or weird because this is a fundamentally different operation than we've done anywhere else. Here, we're dealing with something that could be a "partial failure".

Thinking a bit more on it, though, I don't know that trying to promote everything we can makes any more sense than stopping at the first failure. If we could access the dependency graph of the packages and try and promote as much as we could in _dependency order_, then that seems like a good approach to take, but without that, maybe the code as-is is how we should go.

reset commented:
@thesentinels approve

reset commented:
Part of #2686

thesentinels commented:
:metal: I am testing your branch against master before merging it. We do this to ensure that the master branch is never failing tests.

thesentinels commented:
Sorry; I had a problem testing this pull request. The error was:

```ruby
Expected process to exit with [0], but received '128'
---- Begin output of git fetch origin pull/2866/merge ----
STDOUT: 
STDERR: Warning: Permanently added 'github.com,192.30.255.113' (RSA) to the list of known hosts.
fatal: Couldn't find remote ref pull/2866/merge
---- End output of git fetch origin pull/2866/merge ----
Ran git fetch origin pull/2866/merge returned 128
```

![Oops](https://cloud.githubusercontent.com/assets/4304/17756537/583f4ba4-6495-11e6-97c8-0c22ceaf7e63.gif)

thesentinels commented:
:metal: I am testing your branch against master before merging it. We do this to ensure that the master branch is never failing tests.

thesentinels commented:
Sorry; I had a problem testing this pull request. The error was:

```ruby
Expected process to exit with [0], but received '128'
---- Begin output of git fetch origin pull/2866/merge ----
STDOUT: 
STDERR: Warning: Permanently added 'github.com,192.30.255.112' (RSA) to the list of known hosts.
fatal: Couldn't find remote ref pull/2866/merge
---- End output of git fetch origin pull/2866/merge ----
Ran git fetch origin pull/2866/merge returned 128
```

![Oops](https://cloud.githubusercontent.com/assets/4304/17756537/583f4ba4-6495-11e6-97c8-0c22ceaf7e63.gif)
